### PR TITLE
Adding Logistic Regression class implementation & utils

### PR DESCRIPTION
**Pull request by stompsjo**
This introduces an ML class, `LogReg`, that can be used for supervised logistic regression. This includes typical scikit-learn-esque methods like train and predict as well as methods for hyperparameter optimization and saving the model to file.

### CODE REVIEW

**Comment by stompsjo** [comment]
@gonuke This PR should be reviewed and merged first because it includes `utils.py` which will be used for all subsequent ML PRs (the rest can be done in parallel). This will also be the "most difficult" to review, since it includes both the utils and one ML class (`LogReg`) that reflects the common structure for ML models.

Coveralls says that coverage dropped because the SSML functionality of `utils.cross_validation` are not tested (since this PR does not include any SSML models). A test is already implemented as part of #45.

File: models/LogReg.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    '''
+    Methods for deploying sklearn's logistic regression
+    implementation with hyperparameter optimization.
+    Data agnostic (i.e. user supplied data inputs).
+    TODO: Currently only supports binary classification.
+        Add multinomial functions and unit tests.
+        Add functionality for regression(?)
+    Inputs:
+    params: dictionary of logistic regression input functions.
+        keys max_iter, tol, and C supported.
```
Comment (gonuke) [inline models/LogReg.py:21]: It might make sense to do some checking that `params` has the correct entries.

File: scripts/utils.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+            # if it has been beaten, restart patience
+            self.min_validation_loss = validation_loss
+            self.counter = 0
+        elif validation_loss > (self.min_validation_loss + self.min_delta):
+            # keep track of whether validation loss has been decreasing
+            # by a tolerable amount
+            self.counter += 1
+            if self.counter >= self.patience:
+                return True
+        return False
```
Comment (gonuke) [inline scripts/utils.py]:
```suggestion
        return self.counter >= self.patience
```

File: scripts/utils.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+        results = model.fresh_start(params, data_dict)
+        accs = np.append(accs, results['accuracy'])
+        reports = np.append(reports, results)
+
+    # report cross validation results
+    print('Average accuracy:', np.mean(accs))
+    print('Max accuracy:', np.max(accs))
+    print('All accuracy:', accs)
+    # return the results of fresh_start for the max accuracy model
+    return reports[np.argmax(accs)]
```
Comment (gonuke) [inline scripts/utils.py]: Is the max accuracy split interesting?  The only difference between these is how the data is split, right?

File: scripts/utils.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+
+def pca(Lx, Ly, Ux, Uy, filename):
```
Comment (gonuke) [inline scripts/utils.py]: Separation of concerns - recommend splitting the PCA calculation from the plotter into 2 methods

File: scripts/utils.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    # only saving colors for binary classification with unlabeled instances
+    col_dict = {-1: 'tab:gray', 0: 'tab:orange', 1: 'tab:blue'}
+
+    pcadata = np.append(Lx, Ux, axis=0)
+    normalizer = StandardScaler()
+    x = normalizer.fit_transform(pcadata)
+    print(np.mean(pcadata), np.std(pcadata))
+    print(np.mean(x), np.std(x))
+
+    n = 2
```
Comment (gonuke) [inline scripts/utils.py]: This is an input parameter, but always overwritten here?

File: scripts/utils.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    ax.grid()
+    ax.legend()
+
+    if filename[-4:] != '.png':
+        filename += '.png'
+    fig.tight_layout()
+    fig.savefig(filename)
+
+
+def multiD_pca(Lx, Ly, Ux, Uy, filename, n=2):
```
Comment (gonuke) [inline scripts/utils.py]: There's a lot of room for code re-use in This and previous method

File: scripts/utils.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+                               principalComponents[indices, col],
+                               c=color,
+                               label='class '+str(idx))
+    fig.tight_layout()
+    if filename[-4:] != '.png':
+        filename += '.png'
+    fig.savefig(filename)
+
+
+def plot_cf(testy, predy, title, filename):
```
Comment (gonuke) [inline scripts/utils.py]: Past reviews of these in print suggest that a font size input would be useful 🙂

File: tests/test_models.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+spectra = np.full((len(timestamps), test_data.energy_bins),
+                  np.full((1, test_data.energy_bins), sample_val))
+# setting up for rejected null hypothesis
+rejected_H0_time = np.random.choice(spectra.shape[0],
+                                    test_data.timesteps//2,
+                                    replace=False)
+spectra[rejected_H0_time] = 100.0
+
+labels = np.full((spectra.shape[0],), 0)
+labels[rejected_H0_time] = 1
```
Comment (gonuke) [inline tests/test_models.py]: Seems like this could be a method?

File: tests/test_models.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                                                        test_size=0.2,
+                                                        random_state=0)
+
+    filename = 'test_pca'
+    utils.pca(X_train, y_train, Ux, np.full_like(Uy, -1), filename)
+    os.remove(filename+'.png')
+
+    filename = 'test_multiD_pca'
+    utils.multiD_pca(X_train, y_train, Ux, np.full_like(Uy, -1), filename, n=5)
+    os.remove(filename+'.png')
```
Comment (gonuke) [inline tests/test_models.py]: If you separated the task of performing PCA from plotting, you could possibly avoid testing the plotting.  All you test here is that the plots run successfully.

File: tests/test_models.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    # assert max_acc_model['accuracy'] >= 0.5
+
+    # data split for data visualization
+    X_train, X_test, y_train, y_test = train_test_split(X,
+                                                        y,
+                                                        test_size=0.2,
+                                                        random_state=0)
+
+    filename = 'test_pca'
+    utils.pca(X_train, y_train, Ux, np.full_like(Uy, -1), filename)
```
Comment (gonuke) [inline tests/test_models.py]: Shouldn't this be `x_test`?

File: tests/test_models.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+rejected_H0_time = np.random.choice(spectra.shape[0],
+                                    test_data.timesteps//2,
+                                    replace=False)
+spectra[rejected_H0_time] = 100.0
+
+labels = np.full((spectra.shape[0],), 0)
+labels[rejected_H0_time] = 1
+
+
+def test_utils():
```
Comment (gonuke) [inline tests/test_models.py]: My instinct is to split these into separate tests

File: tests/test_models.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    Uy = np.full_like(Uy, -1)
+
+    # test cross validation for supervised data using LogReg
+    params = {'max_iter': 2022, 'tol': 0.5, 'C': 5.0}
+    model = LogReg(params=params)
+    max_acc_model = utils.cross_validation(model=model,
+                                           X=X,
+                                           y=y,
+                                           params=params)
+    assert max_acc_model['accuracy'] >= 0.5
```
Comment (gonuke) [inline tests/test_models.py]: What's the basis for this test for correctness?

File: tests/test_models.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+    # default behavior
+    model = LogReg(params=None, random_state=0)
+    model.train(X_train, y_train)
+
+    # testing train and predict methods
+    pred, acc = model.predict(X_test, y_test)
+
+    assert acc > 0.7
+    np.testing.assert_equal(pred, y_test)
```
Comment (gonuke) [inline tests/test_models.py]: Why do we expect perfect prediction?

**Review by gonuke — CHANGES_REQUESTED** [review]
This looks correct enough, but a few suggestions for maintainability.

Some good discussion points for a S/W meeting, too.

File: scripts/utils.py (hunk 3 above)
Comment (stompsjo) [reply scripts/utils.py]: The max accuracy split is only interesting in that it has the max accuracy of all the splits tried. K-fold cross validation is useful for either (a) checking for significant skew in accuracy for certain splits or (b) finding the average accuracy over all K splits.

So maybe it is more interesting to return the average accuracy split, or all the results/reports wholesale.

File: tests/test_models.py (hunk 8 above)
Comment (stompsjo) [reply tests/test_models.py]: I've held off making this a method because I sometimes use these variables in my test assertions and therefore want them global for all tests in the file. Not sure if there is a good way to use these variables in a method and share that with all the test functions (maybe with my fixture; _S/W question!_)

File: tests/test_models.py (hunk 10 above)
Comment (stompsjo) [reply tests/test_models.py]: Not necessarily. Since PCA could be used for data dimensionality reduction, it will depend on what the user wants to visualize with PCA. But using `X_test` here would avoid conflating (py)_tests_ with train-_test_-splits.

File: tests/test_models.py (hunk 12 above)
Comment (stompsjo) [reply tests/test_models.py]: In the now defunct #38, I mentioned that I'm having a hard time designing tests for ML models that strike a balance between meaningful data and can be trained in an appropriate amount of time for a unit test. I settled on 0.5 as a goal because if the trained model is achieving an accuracy better than 50% it must be doing _something_. Part of the problem resides in the simplistic data generated above...

A good S/W question (maybe)!

File: tests/test_models.py (hunk 13 above)
Comment (stompsjo) [reply tests/test_models.py]: Good catch, we don't. Line 138 can be incorrect when 137 is correct. The `assert_equal` should probably be removed.

File: tests/test_models.py (hunk 9 above)
Comment (stompsjo) [reply tests/test_models.py]: S/W Question: How should PCA be tested? What is the expected result for a script that essentially relies on the accuracy of [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA.fit_transform)?

File: scripts/utils.py [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+
+    # of all trials, find best and worst loss/accuracy from optimization
+    best = trials.results[np.argmin([r['loss'] for r in trials.results])]
+    worst = trials.results[np.argmax([r['loss'] for r in trials.results])]
+
+    if verbose:
+        logging.info('best accuracy:', 1-best['loss'])
+        logging.info('best params:', best['params'])
+        logging.info('worst accuracy:', 1-worst['loss'])
+        logging.info('worst params:', worst['params'])
```
Comment (stompsjo) [inline scripts/utils.py:95]: Does it make sense to have a verbose setting with logging or should I approach this differently?

File: scripts/utils.py [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+    shuffle: bool; if True, shuffle the data before conducting K-folds.
+    '''
+
+    # return lists
+    accs = []
+    reports = []
+
+    if stratified:
+        cv = StratifiedKFold(n_splits=n_splits, random_state=random_state,
+                             shuffle=shuffle)
```
Comment (stompsjo) [inline scripts/utils.py]: There is currently no test for stratified KFold, but the only difference between this and standard KFold (which is tested) is a different scikit-learn class. I could devise a second test dataset with a few extra instances of one class (which would be balanced by `StratifiedKFold`) or we could ignore testing this portion. Thoughts?

File: scripts/utils.py [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+                else:
+                    ax = _plot_pca_data(principalComponents, Ly, Uy, ax)
+
+    if filename[-4:] != '.png':
+        filename += '.png'
+    fig.tight_layout()
+    fig.savefig(filename)
+
+
+def plot_cf(testy, predy, title, filename):
```
Comment (stompsjo) [inline scripts/utils.py]: @gonuke, per our conversations from S/W, do you still concur that we can ignore testing these PCA/plotting functions?

File: tests/test_models.py [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+    model = LogReg(params=None, random_state=0)
+    model.train(X_train, y_train)
+
+    # testing train and predict methods
+    pred, acc = model.predict(X_test, y_test)
+
+    # since the test data used here is synthetic/toy data (i.e. uninteresting),
+    # the trained model should be at least better than a 50-50 guess
+    # if it was worse, something would be wrong with the ML class
+    assert acc > 0.5
```
Comment (stompsjo) [inline tests/test_models.py:163]: I added some clarification to explain our "good-enough" testing of trained ML models.

**Review by stompsjo — COMMENTED** [review]
@gonuke I have addressed some of the comments from our S/W review. Here are a few lingering comments, pending a re-review from you.

File: scripts/utils.py [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+        results = model.fresh_start(params, data_dict)
+        accs = np.append(accs, results['accuracy'])
+        reports = np.append(reports, results)
+
+    # report cross validation results
+    logging.info('Average accuracy:', np.mean(accs))
+    logging.info('Max accuracy:', np.max(accs))
+    logging.info('All accuracy:', accs)
+    # return the results of fresh_start for the max accuracy model
+    return reports
```
Comment (stompsjo) [inline scripts/utils.py:191]: I changed this to return all the results from each K-fold of cross validation. This was motivated by my new `test_cross_validation`, but I realize it may be valuable to a user to have all information and models produced by this function. A user could just as easily ignore any reports/results that are uninteresting.

File: models/LogReg.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+        # defaults to a fixed value for reproducibility
+        self.random_state = random_state
+        # dictionary of parameters for logistic regression model
+        self.params = params
+        if self.params is None:
+            self.model = linear_model.LogisticRegression(
+                            random_state=self.random_state
+                        )
+        else:
+            if all(key in params.keys() for key in keys):
```
Comment (gonuke) [inline models/LogReg.py]: I'm not sure if this is the most correct/robust way to do this.  The `LogisticRegression` model has defaults for these parameters, so it may be OK if some are missing.  You just need to make sure they exist if you want to pass them along.  Right now, you only allow 0 parameters or all 3 parameters, but maybe it's OK for just 1 or 2?

One way to manage this is with the `**kwargs` object that you can pass through, perhaps?

File: scripts/utils.py (hunk 14 above)
Comment (gonuke) [reply scripts/utils.py:95]: I think the `logging` module gives you a built in way to do this with something like `--log=INFO` on the command-line.

File: scripts/utils.py (hunk 16 above)
Comment (gonuke) [reply scripts/utils.py]: Yes - but is there a way to exclude them from the denominator of the coverage testing?

**Review by gonuke — CHANGES_REQUESTED** [review]
Still some questions about what the right test for parameters passed into the LogReg initializer....

File: scripts/utils.py (hunk 15 above)
Comment (stompsjo) [reply scripts/utils.py]: In light of `# pragma: no cover` from Coveralls, I am adding to this IF...ELSE... since we are not testing StratifiedKFold.

File: scripts/utils.py (hunk 16 above)
Comment (stompsjo) [reply scripts/utils.py]: Turns out, Coveralls supports adding `# pragma: no cover` to blocks of code, and it will ignore it. Adding to the plotting functions above.

File: models/LogReg.py (hunk 19 above)
Comment (stompsjo) [reply models/LogReg.py]: This is my first time using `**kwargs` but I saw a recommendation to use `kwargs.pop('key', default_value)` to pull option args from the input. This system should support any combination of input parameters, including ones that are not supported. I have updated the `__init__` and its relevant unit test. Let me know if you have feedback!

**Comment by stompsjo** [comment]
@gonuke I addressed your comments, thanks!

**Review by gonuke — APPROVED** [review]
Looks good!
